### ISSUE TITLE
Link between Promotion and Adjustment missing

### ISSUE THREAD

**Issue by Nek-** [Sylius/Sylius#14005]
Just like the Shipment on the adjustment, we need a link between adjustment and promotion. This is especially important when it comes to complex calculation with very specific promotion type (which is authorized by the rest of the system and the ability to extends)

**Describe the proposed solution**

```php
namespace Sylius\Component\Core\Model;

use Sylius\Component\Order\Model\Adjustment as BaseAdjustment;

class Adjustment extends BaseAdjustment implements AdjustmentInterface
{
    /**
     * @var ShipmentInterface|null
     */
    protected $shipment; // Note: already existing

    /**
     * @var PromotionInterface|null
     */
    protected $promotion; // Note: the new one
}
```

**Comment by Ferror** [comment]
@Nek- I understand that this could be useful in your use case, but you can simply do it in your codebase, because as you mentioned `complex calculation with very specific promotion type` is probably not a very popular use case.

However, I'm looking forward to getting a little more detailed info about your problem and proposed solution. How do you think your proposal may improve the lives of others? 😃

**Comment by Nek-** [comment]
I think this is a piece of fundamental information missing. It's really like shipment relation, adding this is not trivial while it should actually be part of the model.

**Comment by Ferror** [comment]
Alright. So let me rephrase and ensure I understand it correctly. You state that `PromotionInterface` should be a part of the `Adjustment` model as it is probably the real/best representation of the current behavior?

**Comment by Nek-** [comment]
No, it looks like I was not clear enough. Let me explain a little bit more.

When you order a product with a promotion, the current process is to calculate all adjustments based on the shipment and the promotions. It's fine.

This is how adjustments are generated on the order:
```php
/** @var \Sylius\Component\Promotion\Processor\PromotionProcessorInterface $promotionProcessor */
$promotionProcessor->process($order);
```

After this process, you can calculate the total using only adjustments. You can also know to what shipment the adjustment is linked. **What you cannot know** is on what promotion an adjustment is connected. This may be an issue in many many cases. For example, if I want to list what promotion has been applied to my order, I just can't.

**Comment by Ferror** [comment]
Hi! I just consulted with the Team and we won't add the promotion to the Adjustments, but there is another solution that may help you - you are able to get a promotion code via `$originCode` from Adjustment. However, if you need some more info about an order you can look into the `$details` property. (There is no promotion info at the moment, but you may add them if needed)

**Comment by Nek-** [comment]
originCode looks perfect indeed, missed it! Thanks :) .
